### PR TITLE
Actions

### PR DESCRIPTION
**Pull request by prabalsharma**
Implements https://github.com/cloudfoundry-incubator/kubecf/issues/1144

- [x] should create clusters on GKE
- [x] github actions workflow that runs smoke, CATS, SITS and brain tests for PRs
- [x] runs diego
- [x] runs eirini
- [x] doesn't run for PRs from a fork
- [x] should cleanup any cloud resources
- [x] should not be mandatory for merges
- [x] GKE clusters should have only one node

Test: https://github.com/prabalsharma/kubecf/actions

### CODE REVIEW

File: .github/workflows/pull-request-ci.yaml [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+      with:
+        python-version: '3.x'
+    - name: cache.tools
+      uses: actions/cache@v2
+      with:
+        path: ${{ github.workspace }}/tools
+        key: ${{ runner.os }}-tools
+    - run: make tools-install
+
+    # Run the linters; split them out so it's easier to identify which one failed
```
Comment (viccuad) [inline .github/workflows/pull-request-ci.yaml:30]: Could we use `make lint` to get new linters added to the pipeline for free?
If they don't get enumerated, I would argue that this functionality should be added to `make lint`.

File: .github/workflows/pull-request-ci.yaml [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+        path: ${{ github.workspace }}/catapult/modules/common/bin
+        key: ${{ runner.os }}-catapult-common-tools
+
+    - run: make k8s
+      working-directory: catapult
+      env:
+        GKE_CLUSTER_NAME: kubecf-ci-${{ github.run_id }}-${{ matrix.backend }}
+        OWNER: ${{ github.repository_owner }}
+        DEBUG_MODE: false
+        GKE_NODE_COUNT: 1
```
Comment (viccuad) [inline .github/workflows/pull-request-ci.yaml:172]:
```suggestion
        GKE_NODE_COUNT: 1
        GKE_PREEMPTIBLE: true
```

We should also set the cluster as preemptible, to mirror the concourse setup.

File: .github/workflows/pull-request-ci.yaml [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+      env:
+        GKE_CRED_JSON: ${{ secrets.GKE_CREDS_JSON }}
+
+    - name: cache.catapult-common-tools
+      uses: actions/cache@v2
+      with:
+        path: ${{ github.workspace }}/catapult/modules/common/bin
+        key: ${{ runner.os }}-catapult-common-tools
+
+    - run: make k8s
```
Comment (viccuad) [inline .github/workflows/pull-request-ci.yaml:166]: We could also separate the gke env vars that we are setting in `make k8s` onto it's own task as the `set GKE_CRED_JSON`. That way we could swap between providers easier in the future, just changing the `BACKEND` var.

File: .github/workflows/pull-request-ci.yaml [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+      run:
+        working-directory: kubecf
+
+    strategy:
+      matrix:
+        include:
+        - backend: diego
+          experimental: false
+        - backend: eirini
+          experimental: true
```
Comment (viovanov) [inline .github/workflows/pull-request-ci.yaml:99]: I think we shouldn't consider eirini experimental

File: .github/workflows/pull-request-ci.yaml [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+
+    steps:
+    - name: Checkout KubeCF
```
Comment (viovanov) [inline .github/workflows/pull-request-ci.yaml:102]: Can you please leave one empty row between each step?
If warranted even a comment helps read the yaml.

e.g.

```yaml

    # We need catapult to create Kubernetes clusters
    - name: Checkout catapult
      uses: actions/checkout@v2
      with:
        repository: SUSE/catapult
        path: catapult

    # We'll be caching foo and bar
    - name: cache.tools
      uses: actions/cache@v2
      with:
        path: ${{ github.workspace }}/tools
        key: ${{ runner.os }}-tools
```

File: .github/workflows/pull-request-ci.yaml [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+
+    - id: download-bundle
+      name: Download chart bundle
+      uses: actions/download-artifact@v2
+      with:
+        name: kubecf-bundle.tgz
+        # The artifact upload/download creates an extra directory
+        path: ${{ github.workspace }}/kubecf-bundle
+
+    - name: Download gomplate
```
Comment (viovanov) [inline .github/workflows/pull-request-ci.yaml:130]: This should be a dependency we download using `make <something>`.
@jandubois can help us do that.
Ideally, the code to download dependencies should live in `kubecf-tools`.

File: .github/workflows/pull-request-ci.yaml [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+        if test -x "${TOOLS_DIR}/gomplate" ; then
+          echo "gomplate already exists; skipping."
+          exit 0
+        fi
+        mkdir -p "${TOOLS_DIR}"
+        wget -O "${TOOLS_DIR}/gomplate" \
+          https://github.com/hairyhenderson/gomplate/releases/latest/download/gomplate_linux-amd64
+        chmod a+x "${TOOLS_DIR}/gomplate"
+
+    - name: Start SSH agent
```
Comment (viovanov) [inline .github/workflows/pull-request-ci.yaml:141]: Why do we need this? If for debugging, we should use the tmate action [1].

[1] https://github.com/marketplace/actions/debugging-with-tmate

File: .github/workflows/pull-request-ci.yaml [hunk 8]
```diff
@@ -0,0 +1,307 @@
...
+        set -o errexit -o pipefail -o nounset
+        eval "$(ssh-agent -s)"
+        ssh-keygen -t rsa -b 4096 -N '' -C "KubeCF CI #${{ github.run_id }}" -f ssh-key
+        ssh-add ssh-key
+        echo "::add-mask::$(cat ssh-key.pub)"
+        rm -f ssh-key ssh-key.pub
+        echo "::set-env name=SSH_AUTH_SOCK::${SSH_AUTH_SOCK}"
+        echo "::set-env name=SSH_AGENT_PID::${SSH_AGENT_PID}"
+
+    - name: set GKE_CRED_JSON
```
Comment (viovanov) [inline .github/workflows/pull-request-ci.yaml:152]: Please use the vault action [1] for these secrets with our vault deployment [2].
Please read the guide here on how to use a personal access token to login [3].
The `VAULT_TOKEN` is already available to kubecf.

[1] https://github.com/marketplace/actions/vault-secrets
[2] http://volt.cap.explore.suse.dev/
[3] https://www.vaultproject.io/docs/auth/github#github-auth-method

File: .github/workflows/pull-request-ci.yaml [hunk 9]
```diff
@@ -0,0 +1,307 @@
...
+        # at the end of the run.
+        cat "${KUBECONFIG}" | gzip | base64 --wrap=0
+      working-directory: catapult
+
+    - name: Add tools to path
+      run: |
+        echo "::add-path::${TOOLS_DIR}"
+
+    - name: Generate KubeCF Configuration
+      run: |
```
Comment (viovanov) [inline .github/workflows/pull-request-ci.yaml:191]: We shouldn't generate `values.yaml` anymore.
Do you want to make this change here (seems easier to me), or in a different PR?

File: .github/workflows/pull-request-ci.yaml (hunk 7 above)
Comment (mook-as) [reply .github/workflows/pull-request-ci.yaml:141]: This was for debugging and we can drop it.

The tmate action isn't great because
- It's by some random guy; I've tried to limit the actions here to be first-party Microsoft ones (including GitHub).
- it hangs the CI until you connect and create a file.  That's not useful if we just want to be able to get in only if the run fails or looks suspicious.

File: .github/workflows/pull-request-ci.yaml (hunk 9 above)
Comment (prabalsharma) [reply .github/workflows/pull-request-ci.yaml:191]: Let us handle it in different PR. I think it is more important to get initial commits on GHA which will enable us for quick iterations to improve it.

File: .github/workflows/pull-request-ci.yaml (hunk 7 above)
Comment (prabalsharma) [reply .github/workflows/pull-request-ci.yaml:147]: commented. It is a requirement for catapult.

Comment (viovanov) [reply .github/workflows/pull-request-ci.yaml:147]: @prabalsharma what does catapult use it for?

File: .github/workflows/pull-request-ci.yaml [hunk 10]
```diff
@@ -0,0 +1,319 @@
...
+    name: lint
+    runs-on: ubuntu-latest
+    env:
+      PINNED_TOOLS: true
+      TOOLS_DIR: ${{ github.workspace }}/tools
+    defaults:
+      run:
+        working-directory: kubecf
+    steps:
+    - name: Checkout KubeCF
```
Comment (viovanov) [inline .github/workflows/pull-request-ci.yaml:14]: Can you please add the comment / space for all steps in all jobs?

File: .github/workflows/pull-request-ci.yaml [hunk 11]
```diff
@@ -0,0 +1,307 @@
+name: CI
+on: push
```
Comment (viovanov) [inline .github/workflows/pull-request-ci.yaml:2]: I don't think this does what the description asks for.
Please see [1]. We need this to run for checking PRs, but not those that are opened from forks. Just from local branches.

[1] https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#on

File: .github/workflows/pull-request-ci.yaml (hunk 6 above)
Comment (viovanov) [reply .github/workflows/pull-request-ci.yaml:135]: It's ok not to put it in `kubecf-tools`, but please add a new tool here: https://github.com/cloudfoundry-incubator/kubecf/tree/master/scripts/tools

File: .github/workflows/pull-request-ci.yaml (hunk 9 above)
Comment (viovanov) [reply .github/workflows/pull-request-ci.yaml:203]: Ok, can you please add an issue for it?

File: .github/workflows/pull-request-ci.yaml [hunk 12]
```diff
@@ -0,0 +1,319 @@
...
+
+        {{- $ | toYAML }}
+        EOF
+      env:
+        BACKEND: gke
+        AUTOSCALER: "true"
+
+    - run: make cf-operator-apply
+      env:
+        CF_OPERATOR_URL: ${{ github.workspace }}/kubecf-bundle/cf-operator.tgz
```
Comment (viovanov) [inline .github/workflows/pull-request-ci.yaml:249]: Why is this required?

File: .github/workflows/pull-request-ci.yaml (hunk 7 above)
Comment (prabalsharma) [reply .github/workflows/pull-request-ci.yaml:135]: Sorry, I cannot remember now, I will ask around.

File: .github/workflows/pull-request-ci.yaml (hunk 6 above)
Comment (prabalsharma) [reply .github/workflows/pull-request-ci.yaml]: Done https://github.com/cloudfoundry-incubator/kubecf/pull/1217/commits/9e0ba90cdfe9b72db1ac02f86d916c00002b384e

File: .github/workflows/pull-request-ci.yaml (hunk 8 above)
Comment (prabalsharma) [reply .github/workflows/pull-request-ci.yaml:147]: https://github.com/cloudfoundry-incubator/kubecf/issues/1219

File: .github/workflows/pull-request-ci.yaml (hunk 9 above)
Comment (prabalsharma) [reply .github/workflows/pull-request-ci.yaml:191]: https://github.com/cloudfoundry-incubator/kubecf/issues/1220

File: .github/workflows/pull-request-ci.yaml (hunk 11 above)
Comment (mook-as) [reply .github/workflows/pull-request-ci.yaml:2]: Yep, I did it on push in my branch for spiking; this will need to change before we actually merge.

File: .github/workflows/pull-request-ci.yaml (hunk 7 above)
Comment (mook-as) [reply .github/workflows/pull-request-ci.yaml:174]: https://github.com/SUSE/catapult/wiki/AKS#preparation documents the requirement (remember that I started with AKS).

File: .github/workflows/pull-request-ci.yaml [hunk 13]
```diff
@@ -0,0 +1,361 @@
...
+        path: ${{ github.workspace }}/kubecf-bundle
+
+    # Start SSH agent for catapult
+    - name: Start SSH agent
+      run: |
+        set -o errexit -o pipefail -o nounset
+        eval "$(ssh-agent -s)"
+        ssh-keygen -t rsa -b 4096 -N '' -C "KubeCF CI #${{ github.run_id }}" -f ssh-key
+        ssh-add ssh-key
+        echo "::add-mask::$(cat ssh-key.pub)"
```
Comment (mook-as) [inline .github/workflows/pull-request-ci.yaml:180]: I'm not sure I ever got the mask here to work correctly.

File: .github/workflows/pull-request-ci.yaml [hunk 14]
```diff
@@ -0,0 +1,361 @@
...
+      env:
+        GKE_CLUSTER_NAME: kubecf-ci-${{ github.run_id }}-${{ matrix.backend }}
+        OWNER: ${{ github.repository_owner }}
+        DEBUG_MODE: false
+        GKE_NODE_COUNT: 1
+        GKE_PREEMPTIBLE: true
+        GKE_INSTANCE_TYPE: n1-highcpu-16
+
+    # Providing kubeconfig for debugging
+    - name: export KUBECONFIG
```
Comment (mook-as) [inline .github/workflows/pull-request-ci.yaml:213]: _This_ was the thing I was referring to earlier about debugging.

File: .github/workflows/pull-request-ci.yaml (hunk 12 above)
Comment (mook-as) [reply .github/workflows/pull-request-ci.yaml:273]: Given that we already have it (because we have a bundle), downloading it again seems dumb.  This also ensures we don't accidentally use `latest`.

File: .github/workflows/pull-request-ci.yaml [hunk 15]
```diff
@@ -0,0 +1,361 @@
...
+    # Wait for kubecf pods to be ready
+    - run: make kubecf-wait
+
+    # Run smoke tests
+    - run: make smoke-tests
+
+    # Run brain tests
+    - run: make brain-tests
+
+    # Run sits tests
```
Comment (mook-as) [inline .github/workflows/pull-request-ci.yaml:294]:
```suggestion
    # Run sync integration tests
```
"Run sync integration test suite tests" is… technically valid, but rather verbose.  I'd also be fine with _removing_ the comments for the one-line commands, as they don't actually add any value.

File: .github/workflows/pull-request-ci.yaml (hunk 12 above)
Comment (viovanov) [reply .github/workflows/pull-request-ci.yaml:273]: I would expect `make cf-operator-apply` to install kubecf's dependency.
If it doesn't, it's a bug.

Comment (jandubois) [reply .github/workflows/pull-request-ci.yaml:273]: > I would expect `make cf-operator-apply` to install kubecf's dependency.

It does `helm install cf-operator "$(cf_operator_url)" ...` and `cf_operator_url` will return the resolved URL from `dependencies.yaml`, i.e. `https://s3.amazonaws.com/cf-operators/release/helm-charts/cf-operator-{version}.tgz` with `{version}` replaced by the version also specified in `dependencies.yaml`.

File: .github/workflows/pull-request-ci.yaml [hunk 16]
```diff
@@ -0,0 +1,361 @@
...
+      run: |
+        pwd
+        ls -a
+        echo ${{ github.workspace }}
+        ls -a ${{ github.workspace }}
+        echo $PATH
+        ls -a ${{ github.workspace }}/kubecf/output/bin/
+        echo "::add-path::${{ github.workspace }}/kubecf/output/bin"
+        echo $PATH
+        gomplate -v
```
Comment (mook-as) [inline .github/workflows/pull-request-ci.yaml:162]: `add-path` won't be in effect until _after_ the current step.

File: .github/workflows/pull-request-ci.yaml (hunk 10 above)
Comment (prabalsharma) [reply .github/workflows/pull-request-ci.yaml:21]: done

File: .github/workflows/pull-request-ci.yaml (hunk 11 above)
Comment (prabalsharma) [reply .github/workflows/pull-request-ci.yaml]: done
